### PR TITLE
Fix bang notation in the SLL mode for the most common case

### PR DESCRIPTION
**Pull request by MDoerner**
This PR tries to address the problem that the SLL parser fails to properly resolve bang notation, as noted in issue #4405.

The basic problem is that the SLL parser is not context-aware. Consequently, each subrule matches as much as it can. Unfortunately, an exclamation mark is the type hint for the type `Single`. So, the identifer rule descends into the `typedIdentifier` branch and consumes the exclamation mark still required by the surrounding rule. 

To fix this, this PR introduces a sematic predicate in the `identifier` rule that deactivates the `typedIdentifier` alternative whenever there is an IDENTIFIER token two tokens further down. (A typehint cannot immediately be followed by an identifier.)

Unfortunately, the RHS of the bang notation is an `unrestrictedIdentifier`, which can be more than just an `IDENTIFER`; it allows any keyword and foreign identifiers. Moreover, the LHS identifer can also be a foreign name, in which case it can consit of an arbitrary number of tokens. These edge-cases are not covered by the fix. (Note that predicates must be at the start of an alternative.)

### CODE REVIEW

**Comment by MDoerner** [comment]
Now, bang notation with a foreign name on the RHS is covered as well.

File: Rubberduck.Parsing/Grammar/VBAParser.g4 [hunk 1]
```diff
@@ -579,7 +579,8 @@ subscript : (expression whiteSpace TO whiteSpace)? expression;
 
 unrestrictedIdentifier : identifier | statementKeyword | markerKeyword;
 legalLabelIdentifier : { !(new[]{DOEVENTS,END,CLOSE,ELSE,LOOP,NEXT,RANDOMIZE,REM,RESUME,RETURN,STOP,WEND}).Contains(_input.La(1))}? identifier | markerKeyword;
-identifier : typedIdentifier | untypedIdentifier;
+identifier : {_input.La(3) != IDENTIFIER && _input.La(3) != L_SQUARE_BRACKET || _input.La(1) == L_SQUARE_BRACKET}? typedIdentifier
```
Comment (bclothier) [inline Rubberduck.Parsing/Grammar/VBAParser.g4]: I think this merits a comment explaining the logic because in first part, we look at the third token(?) but in the 2nd part, we look at the first token.

File: RubberduckTests/Grammar/VBAParserTests.cs [hunk 2]
```diff
@@ -2953,6 +2953,50 @@ End Sub
...
+        {
+            const string code = @"
+Sub Test()   
+Dim dict As Scripting.Dictionary
+
+Dim x
+x = dict![a]
+End Sub
+";
+            var parseResult = Parse(code);
```
Comment (bclothier) [inline RubberduckTests/Grammar/VBAParserTests.cs]: Should we also test that the `a` is correctly parsed as an identifier? That may be more important for other cases where we have a expression like `dic![[Test]![Run]]`; the `[Test]![Run]` in this case would be an identifier, not a piece of code.

File: Rubberduck.Parsing/Grammar/VBAParser.g4 (hunk 1 above)
Comment (MDoerner) [reply Rubberduck.Parsing/Grammar/VBAParser.g4]: Males sense; I will add a comment explaining the logic.

File: RubberduckTests/Grammar/VBAParserTests.cs (hunk 2 above)
Comment (MDoerner) [reply RubberduckTests/Grammar/VBAParserTests.cs]: I think that would be a different test and actually be of a different nature than the other tests. None of those test specific parts of the code, just that it parses and that some elements appear the correct number of times. Here, we could test the correct number of `unrestrictedIdentifier`.

BTW, we ne er parse `a` as an identifier: the `unrestrictedIdentifier` in question is `[a]`.

Comment (bclothier) [reply RubberduckTests/Grammar/VBAParserTests.cs]: That is fine by me - I just want to be sure we were covering the supported edge cases correctly. One more thought - do we need a separate test to assert the count of `unrestrictedIdentifer` for a chained bangs expression like `Forms![Main]![SomeControl]`/`Forms!Main!SomeControl` ?

Comment (MDoerner) [reply RubberduckTests/Grammar/VBAParserTests.cs]: I will add a test for chained bang notation without brackets and an ignored one with brackets. The version with brackets on the left cannot work the way we do this right now.

File: Rubberduck.Parsing/Grammar/VBAParser.g4 (hunk 1 above)
Comment (bclothier) [reply Rubberduck.Parsing/Grammar/VBAParser.g4]: Typo corrections...

```
//The predicate in the following rule has been introduced to lessen the problem that VBA uses the same characters used as type hints in other syntactical constructs, 
//e.g. in the bang notation (see withDictionaryAccessExpr). Generally, it is not legal to have an identifier with a opening bracket follow immediately after a type hint.
//The first part of the predicate tries to exclude these two situations. Unfortunately, predicates have to be at the start of a rule. S, an assumption about the number 
//of tokens in the identifier is made. All untypedIdentifers not a foreignNames consist of exactly one token and a typedIdentifier is an untypes one followed by a typeHint,
//again a single token. So, in the majority of situations, the third token is the token following the potential type hint. 
//For foreignNames, no assumption can be made because they consist of a pair of brackets containing arbitrarily many tokens. 
//That is why the second part of the predicate looks at the first character in order to determine whether the identifier is a foreignName.
```
